src/librustc/traits/select.rs

```diff
@@ -719,7 +742,8 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
                 obligation.cause.span,
             ) {
                 Some(obligations) => {
-                    self.evaluate_predicates_recursively(previous_stack, obligations.iter())
+                    self.evaluate_predicates_recursively(previous_stack, obligations.iter(),
```

Comment:
Still, I feel like the *intention*  here was that the predicates we evaluate recursively would have a recursion depth that was incremented relative to `obligation.recursion_depth` -- but there are clearly some cases where that's not true. Perhaps *that* is really the bug?

Reply:
I think you're right - it should be possible to increment the recursion depth of the obligations before passing them to `evaluate_predicates_recursively`, and check that recursion depth at the start of `evaluate_predicates_recursively`.